dbdcnx.c

```diff
@@ -0,0 +1,1120 @@
+        if(box->env->ncset != imp_dbh->ncset) continue;
+        if(box->env->cset != imp_dbh->cset) continue;
+        name = box->buf + 2 + box->pass_len;
+        if(0 != strcmp(name, ctrl->dbname)) continue;
+        name += box->dbname_len + 1;
+        if(0 != strcmp(name, ctrl->uid)) continue;
+        if((int)(box->buf[0]) != (int)(ctrl->pool_rlb)) continue;
+        return box;
+    }
+    return NULL;
+}
+/*}}}*/
+/* new_pool_box{{{*/
+int
+new_pool_box(pTHX_ pool_box_t ** slot, dblogin_info_t * ctrl)
+{
+    imp_dbh_t * imp_dbh = ctrl->imp_dbh;
+    OCIEnv * envhp = imp_dbh->envhp;
+    pool_box_t * box;
+    char * name;
+    OCISPool *poolhp;
+    ub4 mode;
+    ub4 dbname_len = strlen(ctrl->dbname);
+    ub4 user_len = strlen(ctrl->uid);
+    ub4 pwd_len = strlen(ctrl->pwd);
+    size_t boxlen = sizeof(*box) + dbname_len + user_len + pwd_len + 3;
+
+    sword status = OCIHandleAlloc(
+            envhp, (dvoid*)&poolhp, OCI_HTYPE_SPOOL, boxlen, (dvoid**)&box
+    );
+    if (status != OCI_SUCCESS)
+        return local_error(aTHX_ ctrl->dbh, "Failed to allocate OCISPool");
+
+    status = OCIHandleAlloc(envhp, (dvoid*)&box->errhp, OCI_HTYPE_ERROR, 0, NULL);
+    if (status != OCI_SUCCESS)
+    {
+        OCIHandleFree(poolhp, OCI_HTYPE_SPOOL);
+        return local_error(aTHX_ ctrl->dbh, "Failed to allocate OCIError");
+    }
+#ifdef ORA_OCI_112
+    if(ctrl->driver_name_len != 0)
+    {
+        OCIAuthInfo * authp;
+        status = OCIHandleAlloc(envhp, (dvoid*)&authp, OCI_HTYPE_AUTHINFO, 0, NULL);
+        if(status != OCI_SUCCESS)
+        {
+            OCIHandleFree(box->errhp, OCI_HTYPE_ERROR);
+            OCIHandleFree(poolhp, OCI_HTYPE_SPOOL);
+            return local_error(aTHX_ ctrl->dbh, "Failed to allocate OCIAuthInfo");
+        }
+        status = OCIAttrSet(
+            authp, OCI_HTYPE_AUTHINFO,
+            (OraText*)ctrl->driver_name, ctrl->driver_name_len,
+            OCI_ATTR_DRIVER_NAME,
+            box->errhp
+        );
+        if(status != OCI_SUCCESS)
+        {
+            (void)oci_error(ctrl->dbh, box->errhp, status, "OCIAttrSet DriverName");
+            OCIHandleFree(box->errhp, OCI_HTYPE_ERROR);
+            OCIHandleFree(authp, OCI_HTYPE_AUTHINFO);
+            OCIHandleFree(poolhp, OCI_HTYPE_SPOOL);
+            return FALSE;
+        }
+        status = OCIAttrSet(
+            poolhp, OCI_HTYPE_SPOOL,
+            authp, 0,
+            OCI_ATTR_SPOOL_AUTH,
+            box->errhp
+        );
+        if(status != OCI_SUCCESS)
+        {
+            (void)oci_error(ctrl->dbh, box->errhp, status, "OCIAttrSet OCIAuthInfo");
+            OCIHandleFree(box->errhp, OCI_HTYPE_ERROR);
+            OCIHandleFree(authp, OCI_HTYPE_AUTHINFO);
+            OCIHandleFree(poolhp, OCI_HTYPE_SPOOL);
+            return FALSE;
+        }
+    }
+#endif
+
+
+    mode = OCI_SPC_HOMOGENEOUS;
+    if(!ctrl->pool_rlb) mode |= OCI_SPC_NO_RLB;
+    status = OCISessionPoolCreate(
```

Comment:
A side note: after discussions with the OCI library developers, we changed the Python and Node.js drivers to always use threaded mode when a Session Pool is being used. This is done in our common OCI wrapper: https://github.com/oracle/odpi/commit/6815f23433135d2a27995fd8fbfbab462e181d35

Reply:
Sorry, I don't quite understand connection between shown code and threaded mode. Threaded mode is set in dbdimp.c line 550.


```
#if defined(USE_ITHREADS) || defined(MULTIPLICITY) || defined(USE_5005THREADS)                                                                                                                                                                                               
        ctrl.mode |= OCI_THREADED;                                                                                                                                                                                                                                                     
#endif

```

It is defined on OCIEnv. Are you talking about this one, or about something else?